### PR TITLE
Extract metadata data from manifest and update AppVersion

### PR DESCRIPTION
**Pull request by duanxianbo**
[(task:18631)](https://staging.teamyapp.com/teams/107/tasks/18631)

### CODE REVIEW

File: core/dao/sqldb/app_version_change.go [hunk 1]
```diff
@@ -61,6 +62,25 @@ func (a *AppVersionChange) FindAppVersionChangesByAppIDAndVersionNumber(ct conte
 	return NewAppVersionChange().FindAppVersionChangesByAppIDAndVersionNumberWithTx(ct, tx, appID, versionNumber)
 }
 
+func (*AppVersionChange) CreateAppVersionChange(ct context.Context, tx *transaction.Transaction, appVersionChange entity.AppVersionChange) *errs.Error {
+	_, err := tx.SQLTx().ExecContext(
+		ct,
+		`
+		INSERT INTO app_version_change (app_id, version_number, change)
```
Comment (naturetime10) [inline core/dao/sqldb/app_version_change.go]:
```suggestion
		INSERT INTO app_version_change (
		  app_id, 
		  version_number, 
		  change
		) 
```

File: core/dao/sqldb/app_version_price.go [hunk 2]
```diff
@@ -61,6 +61,25 @@ func (a *AppVersionPrice) FindAppVersionPricesByAppIDAndVersionNumber(ct context
 	return a.FindAppVersionPricesByAppIDAndVersionNumberWithTx(ct, tx, appID, versionNumber)
 }
 
+func (*AppVersionPrice) CreateAppVersionPrice(ct context.Context, tx *transaction.Transaction, appVersionPrice entity.AppVersionPrice) *errs.Error {
+	_, err := tx.SQLTx().ExecContext(ct,
+		`
+		INSERT INTO app_version_price (app_id, version_number, currency, amount)
```
Comment (naturetime10) [inline core/dao/sqldb/app_version_price.go]:
```suggestion
		INSERT INTO app_version_price (
		  app_id, 
		  version_number, 
		  currency, 
		  amount
		)
```

File: core/service/app.go [hunk 3]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+	Number         int      `yaml:"number"`
+	AppName        string   `yaml:"app_name"`
+	Description    string   `yaml:"description"`
+	HasUiExtension bool     `yaml:"has_ui_extension"`
+	Changes        []string `yaml:"changes"`
+	Prices         []struct {
+		Currency entity.Currency `yaml:"currency"`
+		Amount   int             `yaml:"amount"`
+	}
+}
```
Comment (naturetime10) [inline core/service/app.go]: Maybe put between line 680 and line line 681?

File: core/service/app.go [hunk 4]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+
+type Uploader func(io.Reader) *errs.Error
```
Comment (naturetime10) [inline core/service/app.go]: Can we put this to the top of this file?

File: core/service/app.go [hunk 5]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+
+/*
+sanitizeTarHeaderName removes the first directory name from the header name.
+The file extracted from tar will have the name of tar as prefix of the file name.
+The prefix is defined by the user
+Eg:
+tar name: app.tar
+before tar: manifest.yaml
+after tar: app/manifest.yaml
+*/
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
/*
    removeTarFilePrefix removes the first directory name from file header name.
    The file extracted from the tar will have the name of tar as prefix in the file name.
    The prefix is defined by the user
    Eg:
        tar name: app.tar
        before tar: manifest.yaml
        after extract from tar: app/manifest.yaml
*/
```

File: core/service/app.go [hunk 6]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+func (a App) sanitizeTarHeaderName(headerName string) string {
+	headerName = tarNameSanitizer.ReplaceAllString(headerName, "")
+
```
Comment (naturetime10) [inline core/service/app.go:649]:
```suggestion
```

File: core/service/app.go [hunk 7]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+/*
+sanitizeTarHeaderName removes the first directory name from the header name.
+The file extracted from tar will have the name of tar as prefix of the file name.
+The prefix is defined by the user
+Eg:
+tar name: app.tar
+before tar: manifest.yaml
+after tar: app/manifest.yaml
+*/
+func (a App) sanitizeTarHeaderName(headerName string) string {
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
func (a App) removeTarFilePrefix(fileHeaderName string) string {
```

File: core/service/app.go [hunk 8]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+	headerName = tarNameSanitizer.ReplaceAllString(headerName, "")
```
Comment (naturetime10) [inline core/service/app.go]: parts := strings.Split(headerName, string(os.PathSeparator))
return filepath.join(parts[1:]...)

File: core/service/app.go [hunk 9]
```diff
@@ -26,9 +29,11 @@ import (
 	"github.com/teamyapp/teamy-backend/core/realtime"
 	"github.com/teamyapp/teamy-backend/core/transaction"
 	"google.golang.org/protobuf/types/known/emptypb"
+	"gopkg.in/yaml.v3"
 )
 
 var appPackageRoot = path.Join("app", "packages")
+var tarNameSanitizer = regexp.MustCompile(`^[^/]+/`)
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
```

File: core/service/app.go [hunk 10]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+		return a.processManifestFile(ct, userID, appID, reader, uploader)
+	default:
+		return uploader(reader)
+	}
+}
+
+func (a App) processManifestFile(ct context.Context, userID uint64, appID uint64, reader *tar.Reader, uploader Uploader) *errs.Error {
+	manifestData := manifest{}
+	multiReader := tmio.NewMultiReaders()
+	readers := multiReader.GenerateMultiReaders(reader, 2)
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
	readers := tmio.NewMultiReaders(reader, 2)
```

Is the additional `GenerateMultiReaders` necessary? Can we merge `NewMultiReaders` and `GenerateMultiReaders` to a single API?

File: core/service/app.go (hunk 4 above)
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
type uploadFunc func(io.Reader) *errs.Error
```

File: core/service/app.go [hunk 11]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+func (a App) sanitizeTarHeaderName(headerName string) string {
+	headerName = tarNameSanitizer.ReplaceAllString(headerName, "")
+
+	return headerName
+}
+
+func (a App) processFile(ct context.Context, userID uint64, appID uint64, reader *tar.Reader, fileName string, uploader Uploader) *errs.Error {
+	switch fileName {
+	case "manifest.yaml":
+		return a.processManifestFile(ct, userID, appID, reader, uploader)
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
		return a.processManifestFile(ct, userID, appID, reader, uploadFunc)
```

File: core/service/app.go [hunk 12]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+before tar: manifest.yaml
+after tar: app/manifest.yaml
+*/
+func (a App) sanitizeTarHeaderName(headerName string) string {
+	headerName = tarNameSanitizer.ReplaceAllString(headerName, "")
+
+	return headerName
+}
+
+func (a App) processFile(ct context.Context, userID uint64, appID uint64, reader *tar.Reader, fileName string, uploader Uploader) *errs.Error {
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
func (a App) processFile(ct context.Context, userID uint64, appID uint64, reader *tar.Reader, fileName string, uploader uploadFunc) *errs.Error {
```

File: core/service/app.go [hunk 13]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+	switch fileName {
+	case "manifest.yaml":
+		return a.processManifestFile(ct, userID, appID, reader, uploader)
+	default:
+		return uploader(reader)
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
		return uploadFunc(reader)
```

File: core/service/app.go [hunk 14]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+		if err != nil {
+			once.Do(func() {
+				wgErr = errs.NewError(errs.InvalidArgument, err.Error())
+			})
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		err := uploader(uploadReader)
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
		err := uploadFunc(uploadReader)
```

File: core/service/app.go [hunk 15]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+		AppID:           appID,
+		CreatedByUserID: userID,
+		Number:          manifestData.Number,
+		AppName:         manifestData.AppName,
+		Description:     manifestData.Description,
+		HasUiExtension:  manifestData.HasUiExtension,
+		IsReady:         true,
+		CreatedAt:       time.Now().UTC(),
+	}
+
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
```

File: core/service/app.go [hunk 16]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+	appVersion := entity.AppVersion{
+		AppID:           appID,
+		CreatedByUserID: userID,
+		Number:          manifestData.Number,
+		AppName:         manifestData.AppName,
+		Description:     manifestData.Description,
+		HasUiExtension:  manifestData.HasUiExtension,
+		IsReady:         true,
```
Comment (naturetime10) [inline core/service/app.go:715]: Are we going to change this to enum in the next PR?

File: core/service/app.go [hunk 17]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+		a.logger,
+		a.transactionFactory,
+		a.stateSyncer,
+		ct,
+	)
+	return txCtx.WithTransactions(false, func(tx *cloudTransaction.Transaction, rtTx *realtime.Transaction) *errs.Error {
+		err := a.appVersionDao.CreateAppVersion(ct, tx, appVersion)
+		if err != nil {
+			return err
+		}
```
Comment (naturetime10) [inline core/service/app.go]: I thought we already created the app version in the UI and we are waiting for the app version to be updated through this logic.

File: core/service/app.go [hunk 18]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+
```
Comment (naturetime10) [inline core/service/app.go:729]:
```suggestion
```

File: core/service/app.go [hunk 19]
```diff
@@ -628,6 +639,133 @@ func (a App) uploadAppPackageFiles(
...
+	wg.Wait()
+	if wgErr != nil {
+		return wgErr
+	}
+
+	appVersion := entity.AppVersion{
+		AppID:           appID,
+		CreatedByUserID: userID,
+		Number:          manifestData.Number,
```
Comment (naturetime10) [inline core/service/app.go]: Maybe we can exclude `versionNumber` from manifest and the `versionNumber` passed into this API?

File: migrations/20230924110050_update-app-version-price-and-change-tables.sql [hunk 20]
```diff
@@ -0,0 +1,27 @@
+-- +migrate Up
+ALTER TABLE app_version_price
```
Comment (naturetime10) [inline migrations/20230924110050_update-app-version-price-and-change-tables.sql:2]: Can we use `app_id, version_number, currency` as PRIMARY KEY instead? There can't be multiple prices in the same currency for the same AppVersion

File: migrations/20230924110050_update-app-version-price-and-change-tables.sql [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+	DROP CONSTRAINT pk_app_version_price;
+
+ALTER TABLE app_version_price
+    ADD CONSTRAINT pk_app_version_price PRIMARY KEY (app_id, version_number, currency, amount);
+
+ALTER TABLE app_version_change
+    DROP CONSTRAINT pk_app_version_change;
+
+ALTER TABLE app_version_change
```
Comment (naturetime10) [inline migrations/20230924110050_update-app-version-price-and-change-tables.sql:12]: Can we add change_id? `change` is string description and is okay to have repetition. It shouldn't be part of primary key.

File: migrations/20230924110050_update-app-version-price-and-change-tables.sql [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+    ADD CONSTRAINT pk_app_version_change PRIMARY KEY (app_id, version_number);
+
+ALTER TABLE app_version_price
+    DROP CONSTRAINT pk_app_version_price;
+
+ALTER TABLE app_version_price
+    ADD CONSTRAINT pk_app_version_price PRIMARY KEY (app_id, version_number);
+
+
```
Comment (naturetime10) [inline migrations/20230924110050_update-app-version-price-and-change-tables.sql]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Comment by naturetime10** [comment]
Great job for getting this feature to work!

**Comment by naturetime10** [comment]
Please rename this PR to `Extract metadata data from manifest and update AppVersion`

File: core/service/app.go [hunk 23]
```diff
@@ -628,6 +638,130 @@ func (a App) uploadAppPackageFiles(
 	return nil
 }
 
+type manifest struct {
+	Number         int      `yaml:"number"`
+	AppName        string   `yaml:"app_name"`
+	Description    string   `yaml:"description"`
+	HasUiExtension bool     `yaml:"has_ui_extension"`
+	Changes        []string `yaml:"changes"`
+	Prices         []struct {
```
Comment (naturetime10) [inline core/service/app.go]: We need to add tag for `prices` also.

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/service/app.go (hunk 23 above)
Comment (duanxianbo) [reply core/service/app.go]: you mean add a new field in app_version_price table?

File: core/service/app.go [hunk 24]
```diff
@@ -628,6 +641,142 @@ func (a App) uploadAppPackageFiles(
...
+type manifest struct {
+	AppName        string   `yaml:"app_name"`
+	Description    string   `yaml:"description"`
+	HasUiExtension bool     `yaml:"has_ui_extension"`
+	Changes        []string `yaml:"changes"`
+	Prices         []struct {
+		Currency entity.Currency `yaml:"currency"`
+		Amount   int             `yaml:"amount"`
+		Tag      string          `yaml:"tag"`
+	}
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
	Prices         []struct {
		Currency entity.Currency `yaml:"currency"`
		Amount   int             `yaml:"amount"`
	} `yaml:"prices"`
```

File: core/service/app.go [hunk 25]
```diff
@@ -628,6 +641,142 @@ func (a App) uploadAppPackageFiles(
...
+}
+
+func (a App) processManifestFile(ct context.Context, userID uint64, appID uint64, versionNumber int, reader *tar.Reader, uploaderFunc UploadFunc) *errs.Error {
+	manifestData := manifest{}
```
Comment (naturetime10) [inline core/service/app.go]: Let inline `manifest` struct into `processManifestFile`.

File: migrations/20230924110050_update-app-version-price-and-change-tables.sql [hunk 26]
```diff
@@ -0,0 +1,33 @@
+-- +migrate Up
+ALTER TABLE app_version_price
+    ADD COLUMN tag VARCHAR(100),
+	DROP CONSTRAINT pk_app_version_price;
```
Comment (naturetime10) [inline migrations/20230924110050_update-app-version-price-and-change-tables.sql]:
```suggestion
    DROP CONSTRAINT pk_app_version_price;
```

File: migrations/20230924110050_update-app-version-price-and-change-tables.sql [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+-- +migrate Down
+DELETE FROM app_version_change;
+
+ALTER TABLE app_version_change
+    DROP CONSTRAINT app_version_change_pkey,
+    DROP COLUMN change_id;
+
+ALTER TABLE app_version_change
+    ADD CONSTRAINT pk_app_version_change PRIMARY KEY (app_id, version_number);
+
```
Comment (naturetime10) [inline migrations/20230924110050_update-app-version-price-and-change-tables.sql:24]:
```suggestion
```

File: migrations/20230924110050_update-app-version-price-and-change-tables.sql [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+
+DELETE FROM app_version_price;
+ALTER TABLE app_version_price
+    DROP CONSTRAINT pk_app_version_price;
+
+ALTER TABLE app_version_price
+    ADD CONSTRAINT pk_app_version_price PRIMARY KEY (app_id, version_number),
+    DROP COLUMN tag;
```
Comment (naturetime10) [inline migrations/20230924110050_update-app-version-price-and-change-tables.sql]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/api/gql/money.go [hunk 29]
```diff
@@ -17,6 +17,10 @@ func (m Money) Amount() int32 {
 	return int32(m.money.Amount)
 }
 
+func (m Money) Tag() string {
+	return m.money.Tag
+}
```
Comment (naturetime10) [inline core/api/gql/money.go]:
```suggestion
```

File: core/entity/money.go [hunk 30]
```diff
@@ -3,4 +3,5 @@ package entity
 type Money struct {
 	Currency Currency
 	Amount   int
+	Tag      string
```
Comment (naturetime10) [inline core/entity/money.go]:
```suggestion
```

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**